apps/prairielearn/src/pages/enroll/enroll.ts

```diff
@@ -64,72 +62,43 @@ router.post(
-      // enrollment limit.
-      if (isEnterprise()) {
-        const { institution, course_instance } = await queryRow(
-          sql.select_course_instance,
-          { course_instance_id: req.body.course_instance_id },
-          z.object({
-            institution: InstitutionSchema,
-            course_instance: CourseInstanceSchema,
-          }),
-        );
+    const { institution, course, course_instance } = await queryRow(
+      sql.select_course_instance,
+      { course_instance_id: req.body.course_instance_id },
+      z.object({
+        institution: InstitutionSchema,
+        course: CourseSchema,
+        course_instance: CourseInstanceSchema,
+      }),
+    );
 
-        const institutionEnrollmentCounts = await getEnrollmentCountsForInstitution({
-          institution_id: institution.id,
-          created_since: '1 year',
-        });
-        const courseInstanceEnrollmentCounts = await getEnrollmentCountsForCourseInstance(
-          req.body.course_instance_id,
-        );
+    const courseDisplayName = `${course.short_name}: ${course.title}, ${course_instance.long_name}`;
 
-        const freeInstitutionEnrollmentCount = institutionEnrollmentCounts.free;
-        const freeCourseInstanceEnrollmentCount = courseInstanceEnrollmentCounts.free;
+    if (req.body.__action === 'enroll') {
+      // Abuse the middleware to authorize the user for the course instance.
+      req.params.course_instance_id = course_instance.id;
+      await promisify(authzCourseOrInstance)(req, res);
 
-        const yearlyEnrollmentLimit = institution.yearly_enrollment_limit;
-        const courseInstanceEnrollmentLimit =
-          course_instance.enrollment_limit ?? institution.course_instance_enrollment_limit;
+      const didEnroll = await insertCheckedEnrollment(res, {
+        institution,
+        course_instance,
+        authz_data: res.locals.authz_data,
+      });
 
-        if (
-          freeInstitutionEnrollmentCount + 1 > yearlyEnrollmentLimit ||
-          freeCourseInstanceEnrollmentCount + 1 > courseInstanceEnrollmentLimit
-        ) {
-          limitExceeded = true;
-        }
+      if (!didEnroll) {
+        // We've already been redirected to the appropriate page; do nothing.
+        return;
       }
 
-      if (!limitExceeded) {
-        // No limits would be exceeded, so we can enroll the user.
-        await queryAsync(sql.enroll, {
-          course_instance_id: req.body.course_instance_id,
-          user_id: res.locals.authn_user.user_id,
-          req_date: res.locals.req_date,
-        });
-        res.redirect(req.originalUrl);
-      } else {
-        // We would exceed an enrollment limit. We won't share any specific
-        // details here. In the future, course staff will be able to check
-        // their enrollment limits for themselves.
-        res.redirect('/pl/enroll/limit_exceeded');
-      }
+      flash('success', `You have added yourself to ${courseDisplayName}.`);
```

Comment:
Flashes! I originally had this as "You have enrolled in...", but all the other language on the page goes out of its way to avoid the word "enroll", presumably so that students don't think this has any effect on their university enrollments. Bikeshed on this language (and than in the removal flash below too).

Reply:
Maybe "join"?

Reply:
I like it! Changed in 4ccf5ba.